File: advent-of-code/day04/python/two.py
import os

input_path = os.path.join(os.path.dirname(__file__), "..", "input.txt")

def total_removable() -> int:
    with open(input_path, "r") as file:
        input = file.readlines()

        rows = len(input)
        columns = len(input[0].strip())
        adjacent_directions = [(-1, -1), (-1, 0), (-1, 1),
                               (0, -1),          (0, 1),
                               (1, -1),  (1, 0),  (1, 1)]

        alive_roll = [[char == "@" for char in line.strip()] for line in input]
        adjacent_roll_count = [[0 for _ in range(columns)] for _ in range(rows)]

        for r in range(rows):
            for c in range(columns):
                if not alive_roll[r][c]:
                    continue
                neighbor_rolls = 0
                for dr, dc in adjacent_directions:
                    rr, cc = r + dr, c + dc
                    if 0 <= rr < rows and 0 <= cc < columns and alive_roll[rr][cc]:
                        neighbor_rolls += 1
                adjacent_roll_count[r][c] = neighbor_rolls

        candidate_queue = []
        queue_head_index = 0
        in_queue = [[False for _ in range(columns)] for _ in range(rows)]

        for r in range(rows):
            for c in range(columns):
                if alive_roll[r][c] and adjacent_roll_count[r][c] < 4:
                    candidate_queue.append((r, c))
                    in_queue[r][c] = True
        
        removed = 0
        while queue_head_index < len(candidate_queue):
            r, c = candidate_queue[queue_head_index]
            queue_head_index += 1

            if not alive_roll[r][c]:
                continue
            if adjacent_roll_count[r][c] >= 4:
                continue

            alive_roll[r][c] = False
            removed += 1

            for dr, dc in adjacent_directions:
                rr, cc = r + dr, c + dc
                if 0 <= rr < rows and 0 <= cc < columns and alive_roll[rr][cc]:
                    adjacent_roll_count[rr][cc] -= 1
                    if adjacent_roll_count[rr][cc] < 4 and not in_queue[rr][cc]:
                        candidate_queue.append((rr, cc))
                        in_queue[rr][cc] = True

        print("Total removable:", removed)
        return removed
    
total_removable()

File: advent-of-code/day04/javascript/two.js
const fs = require("fs");
const path = require("path");

const inputPath = path.join(__dirname, "..", "input.txt");
const input = fs.readFileSync(inputPath, "utf-8").trim().split("\n");

function totalRemovable(lines) {
  const rows = lines.length;
  const columns = rows ? lines[0].length : 0;

  const adjacentDirections = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
  ];

  const aliveRoll = Array.from({ length: rows }, (_, r) =>
    Array.from({ length: columns }, (_, c) => lines[r][c] === "@")
  );

  const adjacentRollCount = Array.from({ length: rows }, () => Array(columns).fill(0));
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < columns; c++) {
      if (!aliveRoll[r][c]) continue;
      let neighborRolls = 0;
      for (const [dr, dc] of adjacentDirections) {
        const rr = r + dr,
          cc = c + dc;
        if (rr >= 0 && rr < rows && cc >= 0 && cc < columns && aliveRoll[rr][cc]) neighborRolls++;
      }
      adjacentRollCount[r][c] = neighborRolls;
    }
  }

  const candidateQueue = [];
  let queueHeadIndex = 0;
  const inQueue = Array.from({ length: rows }, () => Array(columns).fill(false));

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < columns; c++) {
      if (aliveRoll[r][c] && adjacentRollCount[r][c] < 4) {
        candidateQueue.push([r, c]);
        inQueue[r][c] = true;
      }
    }
  }

  let removed = 0;
  while (queueHeadIndex < candidateQueue.length) {
    const [r, c] = candidateQueue[queueHeadIndex++];
    inQueue[r][c] = false;

    if (!aliveRoll[r][c]) continue;
    if (adjacentRollCount[r][c] >= 4) continue;

    aliveRoll[r][c] = false;
    removed++;

    for (const [dr, dc] of adjacentDirections) {
      const rr = r + dr,
        cc = c + dc;
      if (rr >= 0 && rr < rows && cc >= 0 && cc < columns && aliveRoll[rr][cc]) {
        adjacentRollCount[rr][cc]--;
        if (adjacentRollCount[rr][cc] < 4 && !inQueue[rr][cc]) {
          candidateQueue.push([rr, cc]);
          inQueue[rr][cc] = true;
        }
      }
    }
  }

  console.log("Total removable:", removed);
  return removed;
}

totalRemovable(input);
